services/ui_backend_service/api/card.py

```diff
@@ -0,0 +1,207 @@
+        """
+        ---
+        description: Get specific card of a task
+        tags:
+        - Card
+        parameters:
+          - $ref: '#/definitions/Params/Path/flow_id'
+          - $ref: '#/definitions/Params/Path/run_number'
+          - $ref: '#/definitions/Params/Path/step_name'
+          - $ref: '#/definitions/Params/Path/task_id'
+          - $ref: '#/definitions/Params/Builtin/_tags'
+          - $ref: '#/definitions/Params/Builtin/_group'
+          - $ref: '#/definitions/Params/Custom/flow_id'
+          - $ref: '#/definitions/Params/Custom/run_number'
+          - $ref: '#/definitions/Params/Custom/step_name'
+          - $ref: '#/definitions/Params/Custom/task_id'
+          - $ref: '#/definitions/Params/Custom/name'
+          - $ref: '#/definitions/Params/Custom/type'
+          - $ref: '#/definitions/Params/Custom/ds_type'
+          - $ref: '#/definitions/Params/Custom/invalidate'
+          - $ref: '#/definitions/Params/Custom/user_name'
+          - $ref: '#/definitions/Params/Custom/ts_epoch'
+        produces:
+        - application/json
+        responses:
+            "200":
+                description: Returns all artifacts of specified task
+                schema:
+                  $ref: '#/definitions/ResponsesArtifactList'
+            "405":
+                description: invalid HTTP Method
+                schema:
+                  $ref: '#/definitions/ResponsesError405'
+        """
+
+        hash = request.match_info.get("hash")
+        task = await self.get_task_by_request(request)
+        if not task:
+            return web.Response(content_type="text/html", status=404, body="Task not found.")
+
+        cards = await get_cards_for_task(self.cache.artifact_cache, task)
+
+        if hash in cards:
+            return web.Response(content_type="text/html", body=cards[hash]["html"])
+        else:
+            return web.Response(content_type="text/html", status=404, body="Card not found for task.")
+
+
+async def get_cards_for_task(cache_client, task, invalidate_cache=False):
+    pathspec = "{flow_id}/{run_id}/{step_name}/{task_id}".format(
+        flow_id=task.get("flow_id"),
+        run_id=task.get("run_id") or task.get("run_number"),
+        step_name=task.get("step_name"),
+        task_id=task.get("task_name") or task.get("task_id"),
+    )
+    res = await cache_client.cache.GetCards([pathspec], invalidate_cache)
+
+    if res.has_pending_request():
+        async for event in res.stream():
+            if event["type"] == "error":
+                # raise error, there was an exception during fetching.
+                raise CardException(event["message"], event["id"], event["traceback"])
+        await res.wait()  # wait until results are ready
+    data = res.get()
+    if pathspec in data:
+        success, value, detail, trace = unpack_processed_value(data[pathspec])
+        if success:
+            return value
+    # TODO: handle persisted errors with card content processing?
```

Comment:
Can you elaborate what are these "persisted errors"? who persist them and where?

Also, is this a TODO item for this PR?

Reply:
yes the todo was valid for this PR as well. This relates to the cache in general, and is perhaps not applicable to card content as there is no parsing of the values etc. associated, but for any exception that is deemed "unrecoverable" for the backend, we choose to persist the exception in the cache as a value, to skip further executions. 

One example would be calling an action for fetching S3 content with an invalid S3 location.